### PR TITLE
Use U64 as the sole integer type in the core

### LINKED ISSUE #356: Use size type for integers in core

### PR DESCRIPTION
**Pull request by cqc-alec**
Closes #356 .

Also add a `Bit` type in the implementation, needed by some of the quantum stuff. I imagine we will remove this and use `int<1>` from the arithmetic resource once we have implemented the quantum resource.

I expect this to conflict with other open PRs so we may want to think about the most convenient order of merging.

### CODE REVIEW

**Comment by ss2165** [comment]
In #364 for the type argument I've called it `USize`, thoughts?

**Comment by ss2165** [comment]
> Also add a Bit type in the implementation, needed by some of the quantum stuff.

Since we're factoring out the quantum resource shortly anyway, could we just use the 64-bit type for  quantum for the time being?

**Comment by cqc-alec** [comment]
> In #364 for the type argument I've called it `USize`, thoughts?

My first idea was `Size` but I went for `U64` instead because:

* since we have to fix a bit width, we may as well be explicit about it;
* it's also used for some things that are not obviously sizes, at least in examples.

File: specification/hugr.md [hunk 1]
```diff
@@ -1587,7 +1586,7 @@ struct Node{
   ...
 }
 // ((source, offset), (target, offset)
-struct Edge = ((Int, Optional<Int>), (Int, Optional<Int>))
+struct Edge = ((U64, Optional<U64>), (U64, Optional<U64>))
```
Comment (cqc-alec) [inline specification/hugr.md:1589]: Likewise...

**Comment by cqc-alec** [comment]
>     * it's also used for some things that are not obviously sizes, at least in examples.

Unsure about this now. I changed some instances of `Int` in the spec that maybe weren't intended to refer to this type (node indices).

**Comment by cqc-alec** [comment]
Will close this PR and make a new one, as the conflicts are a bit tricky to resolve in a merge.

File: specification/hugr.md [hunk 2]
```diff
@@ -1578,7 +1577,7 @@ struct HUGR {
 
 struct Node{
   // parent node index
-  parent: Int,
+  parent: U64,
```
[the remark this answers is not in the document]
Comment (ss2165) [reply specification/hugr.md:1580]: no these are serialized types - up to the serialization format

**Pull request closed without merging**